idseq_dag/util/command.py

```diff
@@ -186,25 +189,35 @@ def subprocess_scope(*args, **kwargs):
     return wrapper
 
 
-def retry(operation, randgen=random.Random().random):
+def retry(operation, MAX_TRIES=3):
     """Retry decorator for external commands."""
     # Note the use of a separate random generator for retries so transient
     # errors won't perturb other random streams used in the application.
+    invocation = [0]  # a python trick, without which the nested function would not increment a counter defined in the parent
     @wraps(operation)
     def wrapped_operation(*args, **kwargs):
-        remaining_attempts = 3
+        randgen = None
+        remaining_attempts = MAX_TRIES
         delay = 1.0
         while remaining_attempts > 1:
             try:
+                t_start = time.time()
                 return operation(*args, **kwargs)
             except:
-                # Random jitter and exponential delay
-                time.sleep(delay * (1.0 + randgen()))
+                t_end = time.time()
+                if randgen == None:
+                    invocation[0] += 1
+                    randgen = random.Random(os.getpid() * 10000 + invocation[0]).random
```

Comment:
For my own curiosity, I'm confused why `randgen = random.Random().random` doesn't work here. You had noted something about lockstep and drawing from the same random stream, but I'm still confused.

Reply:
Given the copy-on-write fork semantics, if you launch a bunch of subprocesses, they all have identical shadows of the same randgen, seeded with the same seed.   If the problem happens to be that a bunch of these are overloading some service, then all of them will draw the same random numbers from their respective streams, and retry in lockstep, defeating the point of random jitter.   This doesn't occur with threads, where they all share the same generator -- with processes the generator is not actually shared.

Reply:
I don't recommend using processes except when absolutely necessary, and when we first wrote this, we weren't using processes;   but, if you have to use them, this can happen.

Reply:
Makes sense! Thanks for the great explanation.

Reply:
More generally, I think a good rule of thumb here is to never do complex initialization in keyword argument defaults.